CenterForOpenScience/osf.io
# [ENG-4546] [OSF Institutions] Institution Default Storage Region(s) - Phase 1

## Purpose

Implement default storage regions for institution (phase 1): when new users are created via institution SSO, if the institution has set up a pre-selected default region, OSF should set the region as the user’s default region.

- [x] Proof-concept of implementation
- [x] A script/command for adding default regions to institutions
- ~~Postman tests with realCAS~~ Not needed
- ~~Check bAdmin~~ Too slow locally, will verify on staging
- [x] Check limitations

## Changes

* Added a many-to-many relationship between `osf.Institution` and `addons_osfstorage.Region`
  * Implemented via a through-table `InstitutionStorageRegion`
  * Support multiple regions with one preferred choice
* Update SSO authentication class to set default region (if eligible) for new users created via SSO
  * If multiple regions exists for a institution, the preferred one will be set
* Add a command/script to set allowed regions and one preferred region for institutions

## QA Notes

* Staging
  * Verify admin works as expected (list, view and modify institutions)
  * Verify the command work as expected
    * `python3 manage.py set_institution_storage_regions -i <institution_id> -r <region_id> [-p] [-d]`
      * `-p` or `--preferred`: set region as preferred 
      * `-d` or `--dry`: dry-run
      * If there is no existing preferred region (e.g. adding the first region), it will be set as preferred

```
dcrm set_institution_storage_regions -i osftype2 -r us
[osf.management.commands.set_institution_storage_regions]  INFO: Region [us] has been added to Institution [osftype2]
[osf.management.commands.set_institution_storage_regions]  INFO: Region [us] has been set as preferred choice for Institution [osftype2]

dcrm set_institution_storage_regions -i osftype2 -r de -p
[osf.management.commands.set_institution_storage_regions]  INFO: Region [de] has been added to Institution [osftype2]
[osf.management.commands.set_institution_storage_regions]  INFO: Existing preferred regions have been cleared for Institution [osftype2]
[osf.management.commands.set_institution_storage_regions]  INFO: Region [de] has been set as preferred choice for Institution [osftype2]

dcrm set_institution_storage_regions -i osftype2 -r us
[osf.management.commands.set_institution_storage_regions]  WARNING: Region [us] already set for Institution [osftype2]

dcrm set_institution_storage_regions -i osftype2 -r us -p
[osf.management.commands.set_institution_storage_regions]  WARNING: Region [us] already set for Institution [osftype2]
[osf.management.commands.set_institution_storage_regions]  INFO: Existing preferred regions have been cleared for Institution [osftype2]
[osf.management.commands.set_institution_storage_regions]  INFO: Region [us] has been set as preferred choice for Institution [osftype2]

dcrm set_institution_storage_regions -i osftype2 -r us -p
[osf.management.commands.set_institution_storage_regions]  WARNING: Region [us] already set for Institution [osftype2]
[osf.management.commands.set_institution_storage_regions]  WARNING: Region [us] already set as preferred for Institution [osftype2]
```

* Prod
  * In addition to staging notes, verify SSO work as expected

## Documentation

N/A

## Side Effects

~~It is possible to create more than one preferred regions for one institution since there is no DB constraint if the provided command is not used. In addition, using the command will fix/clear extra preferred regions when setting a new preferred region.~~ N/A

## Ticket

https://openscience.atlassian.net/browse/ENG-4546
